DTensor: avoiding crashing on dynamic shapes in a few places

bdhirsh opened the pull request:
This was the result of me slogging through errors from running this repro: https://github.com/pytorch/torchtrain/issues/61.

This doesn't get dynamic shapes fulling working, but it does fix some obvious errors I hit.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123349
* #123350
* #124401
* #124400
* #124399
* #124398



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

bdhirsh committed:
DTensor: avoiding crashing on dynamic shapes in a few places

[ghstack-poisoned]

bdhirsh committed:
Update on "DTensor: avoiding crashing on dynamic shapes in a few places"




[ghstack-poisoned]

zou3519 commented on torch/distributed/_tensor/api.py:
```diff
@@ -293,6 +299,8 @@ def __coerce_same_metadata_as_tangent__(self, metadata_tensor):
     # pyre-fixme[3]: Return type must be annotated.
     # pyre-fixme[2]: Parameter must be annotated.
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        if func in metadata_handlers:
+            return metadata_handlers[func](*args, **kwargs)
```
What are the chances this could be handled in @register_sharding instead?

bdhirsh replied:
The main reason I didn't try is because I saw that everything registered via `register_sharding_prop_rule` needs to return an `OutputSharding` ([link](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/sharding_prop.py#L49)), which seems to expect the output of each op to be a tensor or list of tensors ([link](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/op_schema.py#L23)).

We could probably make the logic that invokes the sharding rules more generic so that they can be used to register ops that don't return tensors (cc @wanchaol). Although for metadata ops case, I figured that sharding rules vs. metadata requests were different enough that reasoning about them would be simpler if they were separate (open to either option though)

wanchaol replied:
I think the metadata handling is sorta special that we don't need to make them go through sharding propagation, suggested in another comment to make them custom op handlers

wanchaol reviewed (commented):
Have some suggestions and questions inlined

ezyang commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -406,6 +406,16 @@ def __post_init__(self):
         self._hash: Optional[int] = None
+        if self.tensor_meta is not None:
+            size_symints = any(
+                isinstance(x, torch.SymInt) for x in self.tensor_meta.shape
```
You can't hash SymInts unless you're willing to put a lot of guards on them.

Specifically, suppose you have `hash(s0) == hash(s1)`. If you hash SymInt using the sympy expression, this evaluates to False, but when the runtime values s0 == s1, then the expression should have been True.

wanchaol replied:
I'm a bit confusing here (or maybe lack of knowledge on sympy): if we have `hash(s0) == hash(s1)` then:

> If you hash SymInt using the sympy expression, this evaluates to False

How would this evaluates to False instead of true 🤔

ezyang commented:
Should we have a discussion about this PR?

bdhirsh commented:
Yep, happy to. This PR was the result of “hack through dynamic shapes bugs when running torch train”. But I think more broadly:

(1) (unrelated to this PR)  shape + subclass handling needs some more work. The only remaining “major” bug I’m aware of is that we need to generate an FX graph of the shape compute to properly compute tensor subclass output shapes. But there is generally very little testing for subclasses + dynamic shapes so far

(2) DTensor heavily relies on caching for eager perf. My strategy so far has been to spot check places that DTensor caches, and try to disable them when we see dynamic shapes (really what we probably want is to disable caching when we are in the middle of compiling. I don’t think we have a great way to detect this today, but we could add one)

bdhirsh committed:
Update on "DTensor: avoiding crashing on dynamic shapes in a few places"

This was the result of me slogging through errors from running this repro: https://github.com/pytorch/torchtrain/issues/61.

This doesn't get dynamic shapes fulling working, but it does fix some obvious errors I hit.




[ghstack-poisoned]

wanchaol commented:
Yeah it would be great if we can have some discussion on this!

bdhirsh committed:
Update on "DTensor: avoiding crashing on dynamic shapes in a few places"

This was the result of me slogging through errors from running this repro: https://github.com/pytorch/torchtrain/issues/61.

This doesn't get dynamic shapes fulling working, but it does fix some obvious errors I hit.




[ghstack-poisoned]

bdhirsh committed:
Update on "DTensor: avoiding crashing on dynamic shapes in a few places"

This was the result of me slogging through errors from running this repro: https://github.com/pytorch/torchtrain/issues/61.

This doesn't get dynamic shapes fulling working, but it does fix some obvious errors I hit.




[ghstack-poisoned]
